### ISSUE TITLE
feat: MODEXP

### ISSUE THREAD

**Issue by ClementWalter** [kkrt-labs/keth#177]
## What

Add the modexp precompile

## How

- Check the existing `modexp.cairo` file and test
- use it in `precompiles.cairo` or dev

**Comment by enitrat** [comment]
@ClementWalter this probably needs to use GaragaZero ?

**Comment by enitrat** [comment]
see https://github.com/feltroidprime/garaga-zero/blob/main/src/modulo_circuit.cairo

**Comment by ClementWalter** [comment]
no plan yet, needs to be scoped

**Comment by zmalatrax** [comment]
Scope:
- Need to select a fast exponentiation algorithm.
  - Fast exponentiation by squaring: [wikipedia article](https://en.wikipedia.org/wiki/Exponentiation_by_squaring)
  - Some [considerations](https://en.wikipedia.org/wiki/Modular_exponentiation) for mod exp (to be more memory/time-efficient as we keep the intermediate results within mod p)
  - The easiest to implement is a basic method based on left-to-right binary exponentiation (or right-to-left) (O(log(n))
  - Most cryptographic systems use the [Montgomerry multiplication](https://en.wikipedia.org/wiki/Montgomery_modular_multiplication) for their modular exponentiation algorithms
    - normal basis > Montgomerry basis (perform exponentiation with Montgomerry multiplication) >  normal basis
    - Inconvenient of Montgomerry is that it requires the GCD
> Many important cryptosystems such as [RSA](https://en.wikipedia.org/wiki/RSA_(cryptosystem)) and [Diffie–Hellman key exchange](https://en.wikipedia.org/wiki/Diffie%E2%80%93Hellman_key_exchange) are based on arithmetic operations modulo a large odd number, and for these cryptosystems, computations using Montgomery multiplication with R a power of two are faster than the available alternatives.

See 14.3.2 and 14.6 from the [Handbook of applied cryptography](https://cacr.uwaterloo.ca/hac/about/chap14.pdf).

All algorithms for fast exponentiation seem to require **recursion/loops**, we'll need to update the circuit compiler to handle jump instructions then

**Comment by ClementWalter** [comment]
control flow needs to be kept in regular cairo. I'm not an expert of these algo but it seems like fast exp. by squarring, and using `mul_mod` for the squarring, in a regular cairo 0 function, is the easiest way to start with

**Comment by zmalatrax** [comment]
Indeed, we'll have a `mod_exp` function in Cairo, which would implement the square-and-multiply algo using the `mul` circuit from `mod_ops_compiled.cairo`

I was thinking of doing a mod_exp circuit, but it actually wasn't making much sense..

Later on we could see if implementing the Montgomery exponentiation yields better performance, very low prio though

**Comment by zmalatrax** [comment]
>  in cairo1 we had a modexp implementation using miller loops and ec_mul. have you checked?

Isn't miller loop on elliptic curve points, for pairing ? modexp is modular exponentiation on integers afaik

**Comment by enitrat** [comment]
yeah you're right - it was not miller loops but using the square-and-multiply algorithm. see https://github.com/kkrt-labs/kakarot/pull/1513/files#diff-6c19d3f0ca0bde364eca6e7a950d366792b3324be5dc4221eb22a8676979c100

**Comment by zmalatrax** [comment]
Oh okay, then we'll be implementing the same algorithm - good to know, thanks
